### PR TITLE
[test][sagemaker]Sm remote smart retry

### PR DESCRIPTION
**Pull request by SergTogul**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/testrunner.py [hunk 1]
```diff
@@ -187,8 +187,15 @@ def run_sagemaker_remote_tests(images, pytest_cache_params):
...
             return
         pool_number = len(images)
-        with Pool(pool_number) as p:
-            p.starmap(sm_utils.execute_sagemaker_remote_tests, [[image, pytest_cache_params] for image in images])
+        # Using Manager().dict() since it's a tread safe dictionary
+        global_pytest_cache = Manager().dict()
+        try:
+            with Pool(pool_number) as p:
+                p.starmap(sm_utils.execute_sagemaker_remote_tests,
+                                  [[i, images[i], global_pytest_cache, pytest_cache_params] for i in range(0, pool_number)])
```
Comment (saimidu) [inline test/testrunner.py:195]: Using `range(0, x)` is equivalent to `range(x)`.

Also please correct the indentation to:
```suggestion
                p.starmap(
                    sm_utils.execute_sagemaker_remote_tests,
                    [[i, images[i], global_pytest_cache, pytest_cache_params] for i in range(pool_number)]
                )
```

File: test/test_utils/pytest_cache.py [hunk 2]
```diff
@@ -133,6 +138,32 @@ def upload_pytest_cache_from_local_to_s3(self,
...
+        :param build_context
+        :param test_type
+        """
+        if len(cache_json) == 0:
+            LOGGER.info("No cache was generated. Skip uploading.")
+            return
+        s3_file_dir = self.__make_s3_path(commit_id, framework, version, build_context, test_type)
+        s3_file_path = os.path.join(s3_file_dir, "lastfailed")
+        with open("tmp_file_for_cache_json", "w") as f:
+            f.write(json.dumps({**cache_json}))
```
Comment (saimidu) [inline test/test_utils/pytest_cache.py:164]: Please use `json.dump(<object>, f)` rather than using `f.write(json.dumps(<object>))`.

Also, please use `dict(cache_json)` to convert it from `DictProxy` to `dict`, rather than `{**cache_json}`, which is a roundabout and slightly confusing way of doing the same.

```suggestion
            json.dump(dict(cache_json), f)
```

File: test/test_utils/pytest_cache.py [hunk 3]
```diff
@@ -160,6 +193,16 @@ def merge_cache_and_upload_from_local_to_s3(self,
         s3_file_path = os.path.join(s3_file_dir, "lastfailed")
```
[the remark this answers is not in the document]
Comment (SergTogul) [reply test/test_utils/pytest_cache.py:191]: It's just a buffer accumulating results from different executions. Used only in one function

**Review by tejaschumbalkar — APPROVED** [review]

**Review by tejaschumbalkar — CHANGES_REQUESTED** [review]
LGTM, lets revert temp configs

**Review by tejaschumbalkar — APPROVED** [review]
